File: process.py

# The Process class represents a process in our system.

# When we print processes, we will print them in json format
import json

class Process:

    # Contructor - the default values for the member variables
    # are as indicated - an empty string for the process name,
    # and -1 for the process id
    def __init__(self, process_name="", process_id=-1, priority = 0, cpu_needed = 0):
        self.process_name = process_name
        self.process_id = process_id
        self.my_priority = priority
        self.cpu_needed = cpu_needed
        self.cpu_used = 0

    
    # Set the process name 
    def setProcessName(self, process_name):
        self.process_name = process_name

    # Set the process ID
    def setProcessId(self, process_id):
        self.process_id = process_id

    # Set the prioroty
    def setPriority(self, priority):
        self.my_priority = priority

    def setCpuNeeded(self, cpu_needed):
        self.cpu_needed = cpu_needed

    # get the process name
    def processName(self):
        return self.process_name
    
    # get the process id
    def processId(self):
        return self.process_id

    def priority(self):
        return self.my_priority

    def cpuNeeded(self):
        return self.cpu_needed

    def cpuUsed(self):
        return self.cpu_used

    def run(self, num_cycles):
        cycles_to_run = min(num_cycles, self.cpu_needed - self.cpu_used)

        self.cpu_used += cycles_to_run
        return cycles_to_run


    # This method is called when print() is called. 
    # It converts the contents of the object
    # to json, and then dumps the json format. 
    def __str__(self):
        return json.dumps({"process_name" : self.process_name,
                           "process_id" : self.process_id,
                           "priority" : self.my_priority,
                           "cpu_needed" : self.cpu_needed,
                           "cpu_used" : self.cpu_used}, 
                           separators = (",",":"), sort_keys = True)

File: process.cpp
// process.cpp
//
// Include your implementation of the Process class here.

#include <string>
#include <iostream>
#include <sstream>

#include "process.h"
#include "nlohmann/json.hpp"

using namespace std;

Process::Process(const string& processName, int priority, int cpuNeeded, int processId) :
    m_processName(processName), 
    m_priority(priority), 
    m_cpuNeeded(cpuNeeded), 
    m_processId(processId), 
    m_cpuUsed(0)
{
    // Nothing to do here
}

const string& Process::processName() const
{
    return m_processName;
}

void Process::setProcessName(const string& processName)
{
    m_processName = processName;
}

const int Process::processId() const
{
    return m_processId;
}

void Process::setProcessId(int processId)
{
    m_processId = processId;
}


const int Process::priority() const
{
    return m_priority;
}

void Process::setPriority(int priority)
{
    m_priority = priority;
}


const int Process::cpuNeeded() const
{
    return m_cpuNeeded;
}


void Process::setCpuNeeded(int cpuNeeded)
{
    m_cpuNeeded = cpuNeeded;
}

const int Process::cpuUsed() const
{
    return m_cpuUsed;
}


int Process::run(int numCycles)
{    
    int toRun;

    if ((m_cpuNeeded - m_cpuUsed) >= numCycles) {
        toRun = numCycles;
    } else {
        toRun = m_cpuNeeded - m_cpuUsed;
    }

    m_cpuUsed += toRun;
    return toRun;
}



// Friend function to print the object
std::ostream& operator<<(std::ostream& os, const Process& p)
{
    nlohmann::json j;
    to_json(j, p);
    os << j;
    return os;
}

// Functions to convert a process to and from json
void to_json(nlohmann::json& j, const Process& p)
{
    j["process_name"] = p.processName();
    j["process_id"] = p.processId();
    j["priority"] = p.priority();
    j["cpu_needed"] = p.cpuNeeded();
    j["cpu_used"] = p.cpuUsed();
}

void from_json(const nlohmann::json& j, Process& p)
{
    if (j.contains("process_name")) {
        p.setProcessName(j["process_name"]);
    } else {
        ostringstream o;
        o << "Missing process_name in json input: " << j;
        throw(std::invalid_argument(o.str()));
    }

    if (j.contains("process_id")) {
        p.setProcessId(j["process_id"]);
    }

    if (j.contains("priority")) {
        p.setPriority(j["priority"]);
    }

    if (j.contains("cpu_needed")) {
        p.setCpuNeeded(j["cpu_needed"]);
    }
}